sphinx/ext/autodoc/__init__.py

```diff
@@ -391,11 +391,15 @@ def format_signature(self, **kwargs: Any) -> str:
         else:
             # try to introspect the signature
             try:
-                try:
-                    args = self.format_args(**kwargs)
-                except TypeError:
-                    # retry without arguments for old documenters
-                    args = self.format_args()
+                if kwargs:
+                    try:
+                        args = self.format_args(**kwargs)
+                    except TypeError:
+                        # avoid chaining tracebacks, by putting nothing here
+                        pass
+
+                # retry without arguments for old documenters
+                args = self.format_args()
```

Comment:
It was called twice before too - this way, we at least:
* Do not call it twice with identical arguments
* Do not chain the exception of the first call with the exception of the second

Does anything actually use these arguments?

Reply:
>It was called twice before too

It's an unexpected result. The `TypeError` handler is added only for old Documener classes which does not accept keyword arguments. It implicitly expects `format_args()` method does not raise `TypeError`.

In this case, I think no reason to raise `TypeError` on `format_args()`. It should be handled inside it.

>Does anything actually pass non-empty kwargs?

Yes. The autosummary extension gives keyword argument.
https://github.com/sphinx-doc/sphinx/blob/62711bc081744cecba2af799ad8dbf678f6a97a4/sphinx/ext/autosummary/__init__.py#L345



Note: It's after 2 AM in Japan. So this is my last comment today. Good night!